cpp/ql/test/query-tests/Critical/MemoryFreed/MemoryNeverFreed.expected

```diff
@@ -6,11 +6,16 @@
 | my_auto_ptr.cpp:82:9:82:15 | new | This memory is never freed. |
 | test.cpp:26:18:26:23 | call to malloc | This memory is never freed. |
 | test.cpp:42:18:42:23 | call to malloc | This memory is never freed. |
+| test.cpp:71:18:71:23 | call to malloc | This memory is never freed. |
 | test.cpp:73:18:73:23 | call to malloc | This memory is never freed. |
+| test.cpp:88:18:88:23 | call to malloc | This memory is never freed. |
```

Comment:
These are false positive cases where memory is allocated in a class (constructor or other method) and freed in the destructor of the same class.  These will be extremely common in real world databases and we should have a strategy to deal with them.

Reply:
@MathiasVP I should acknowledge that I'm pretty much just echoing what you said in the meeting here.

Reply:
What's not clear to me is whether it's a matter of finding a solution for destructors, or whether there will be a much wider range of cases where we don't get flow and thus get false positive results.

Reply:
Indeed. In the meeting where just had I expressed the opinion that I don't think raising an alert on all allocations where we cannot find a path to a deallocation will give us a sufficiently low FP rate. For example, since dataflow doesn't handle a lot of aliasing, so at any point if an aliased pointer is free'd we'd miss that path and report a FP.

Instead, how about we create an IR query that tracks flow from an `InitializeDynamicAllocation` that isn't conflated with all-aliased-memory, and find all the ones that don't flow to a deallocation.

Since these aren't all-aliased-memory, the IR's "mostly sound" alias analysis should ensure that we don't miss any potential aliases going into a deallocation, right? So if we have a non-`all-aliased-memory` memory produced by an `InitializeDynamicAllocation` and don't manage to find any flow to a deallocation function, then the "mostly soundness" of the IR should give us a good result, no?

Obviously, since the IR's alias analysis conflates stuff into all-aliased-memory quite quickly, this will give us quite a lot of FNs, but it will hopefully produce mostly good TPs for the results we do produce, right? 🤞

Reply:
> or whether there will be a much wider range of cases where we don't get flow and thus get false positive results.

I have a feeling we'll get a lot of FPs from missing flow. The query as-is will basically turn most "missing flow" into a FP. We've tried that in the past with `cpp/invalid-pointer-deref` (which turns many "missing range analysis" into a query FP). It took many months to get it to a medium precision.

Reply:
Yeah, I agree we should aim to exclude results where we hit something we're not sure about, to avoid FPs.  There should be plenty of good results left for a query like this.  If we can identify the locations where aliasing occurs we can just treat those as probably-frees, i.e. make them sinks in the `AllocationDeallocationConfig` config.